### PR TITLE
Record the correct target type when coercing fn items/closures to pointers

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory. We were previously not recording the *target* type of a coercion as the output of an adjustment. This should remedy that.

We must also modify the function pointer casts in MIR typeck to use subtyping, since those broke since #118247.

r? lcnr

### CODE REVIEW

File: tests/mir-opt/build_correct_coerce.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// skip-filecheck
+
+fn foo<'a, 'b>(_: &'a (), _: &'b ()) {}
```
Comment (lcnr) [inline tests/mir-opt/build_correct_coerce.rs:3]: please add an explanation of what we want to see here

**Review by lcnr — COMMENTED** [review]
can you also add a test which requires the `sub` for closures, is case nothing hit that before?

**Comment by compiler-errors** [comment]
What the heck -- this PR caused two regressions: 

* [Struggleton.TitleScreenBGM.23bd2077802ae42336a788678a462d067fc9aa20](https://crater-reports.s3.amazonaws.com/pr-129021-1/try%23f781a78d815e6a304bf6061b8a8fa69d7ce8dfd5/gh/Struggleton.TitleScreenBGM/log.txt)

```
[INFO] [stdout] error[E0308]: mismatched types
[INFO] [stdout]   --> src/lib.rs:30:4
[INFO] [stdout]    |
[INFO] [stdout] 30 | fn create_press_any_hook(ctx: &mut InlineCtx) {
[INFO] [stdout]    |    ^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
[INFO] [stdout]    |
[INFO] [stdout]    = note: expected fn pointer `unsafe extern "C" fn(&mut InlineCtx)`
[INFO] [stdout]               found fn pointer `unsafe extern "C" for<'a> fn(&'a mut InlineCtx)`
```

* [kwap-0.10.0](https://crater-reports.s3.amazonaws.com/pr-129021-1/try%23f781a78d815e6a304bf6061b8a8fa69d7ce8dfd5/reg/kwap-0.10.0/log.txt)

```
[INFO] [stdout] error[E0282]: type annotations needed
[INFO] [stdout]   --> src/blocking/server.rs:98:8
[INFO] [stdout]    |
[INFO] [stdout] 45 | #[derive(Clone, Debug)]
[INFO] [stdout]    |          ----- in this derive macro expansion
[INFO] [stdout] ...
[INFO] [stdout] 98 |   Send(Addrd<platform::Message<P>>),
[INFO] [stdout]    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for reference `&net::Addrd<kwap_msg::Message<<P as Platform>::MessagePayload, _, <P as Platform>::MessageOptions>>`
[INFO] [stdout]    |
[INFO] [stdout]    = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I'll look into both of them, and revert this if not.

**Pull request merged**